src/zenml/integrations/modal/step_operators/modal_step_operator.py

```diff
@@ -0,0 +1,208 @@
+        entrypoint_command: List[str],
+        environment: Dict[str, str],
+    ) -> None:
+        """Launch a step run on Modal.
+
+        Args:
+            info: The step run information.
+            entrypoint_command: The entrypoint command for the step.
+            environment: The environment variables for the step.
+
+        Raises:
+            SystemExit: If the modal app exits with a non-zero code.
+            ValueError: If no Docker credentials are found for the container registry.
+        """
+        settings = cast(ModalStepOperatorSettings, self.get_settings(info))
+        image_name = info.get_image(key=MODAL_STEP_OPERATOR_DOCKER_IMAGE_KEY)
+        zc = Client()
+        stack = zc.active_stack
+
+        if not stack.container_registry:
+            raise ValueError(
+                "No Container registry found in the stack. "
+                "Please add a container registry and ensure "
+                "it is correctly configured."
+            )
+
+        if docker_creds := stack.container_registry.credentials:
+            docker_username, docker_password = docker_creds
+        else:
+            raise ValueError(
+                "No Docker credentials found for the container registry."
+            )
+
+        # get the pydantic version in local environment
+        # use it to install the correct version of pydantic in the modal app
+        # since it overwrites with 1.x on top of our image
+        pydantic_version = pkg_resources.get_distribution("pydantic").version
+        major, minor, *_ = pydantic_version.split(".")
+
+        # Construct the decorator arguments based on the settings
+        decorator_args = []
+        if settings.gpu is not None:
+            decorator_args.append(f"gpu='{settings.gpu}'")
+        if settings.cpu is not None:
+            decorator_args.append(f"cpu={settings.cpu}")
+        if settings.memory is not None:
+            decorator_args.append(f"memory={settings.memory}")
+        if settings.region is not None:
+            decorator_args.append(f"region='{settings.region}'")
+
+        decorator_args_str = ", ".join(decorator_args)
+
+        with tempfile.NamedTemporaryFile(
+            mode="w", delete=False, suffix=".py"
+        ) as tmp:
+            modal_code = f"""import modal
+
+my_secret = modal.Secret.from_dict({{
+    'REGISTRY_USERNAME': '{docker_username}',
+    'REGISTRY_PASSWORD': '{docker_password}',
+}})
+
+zenml_image = modal.Image.from_registry(tag='{image_name}', secret=my_secret).env({environment}).pip_install("pydantic~={major}.{minor}").run_commands('{" ".join(entrypoint_command)}')
+
+app = modal.App('{info.run_name}')
+
+@app.function(image=zenml_image, {decorator_args_str})
+def run_step():
+    print("Executing {info.run_name} step...")
+
+if __name__ == "__main__":
+    run_step()
+"""
+            tmp.write(modal_code)
+            tmp_filename = tmp.name
+
+        # Create a Click context object
+        ctx = click.Context(run)
```

Comment:
I would suggest using normal subprocess calls instead of click for this. Once we implement a slimmer client version that will be installed in the docker images, click will not be part of that package which would lead to issues here.